### PR TITLE
Implement visitor for toExpression functions

### PR DESCRIPTION
**Pull request by RazvanN7**
Related PRs:

- [x] #7031 
- [x] #7007 
- [ ] #7048

### CODE REVIEW

**Comment by UplinkCoder** [comment]
You added a new file which now contains semantically unrelated code in close visual proximity.
And perf slow downs.

**Comment by WalterBright** [comment]
The `toExpression()` for `type` and `toExpression()` for `Initializer` should be named `typeToExpression()` and `initializerToExpression()` as they have nothing to do with each other (and hence should be distinguished by name rather than overloading). I recommend `typeToExpression()` going into `initsem.d`, and `typeToExpression()` into `typesem.d`.

**Comment by RazvanN7** [comment]
I love it how the autotester passes all tests and the rest are red

**Comment by RazvanN7** [comment]
@WalterBright done. Is it ok now?

**Comment by RazvanN7** [comment]
I thought about aliasing the functions to toExpression, but it is faster to call `initializerToExpression` and `typeToExpression` directly.

**Comment by andralex** [comment]
Ran speed measurements in the same manner as https://github.com/dlang/dmd/pull/7031. Baseline (master):

```
make -j1 -f posix.mak > /dev/null  5.01s user 0.16s system 91% cpu 5.674 total
make -j1 -f posix.mak > /dev/null  4.97s user 0.15s system 93% cpu 5.494 total
make -j1 -f posix.mak > /dev/null  5.12s user 0.06s system 93% cpu 5.550 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.09s system 93% cpu 5.477 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.14s system 93% cpu 5.504 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.13s system 93% cpu 5.521 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.14s system 93% cpu 5.516 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.16s system 93% cpu 5.483 total
make -j1 -f posix.mak > /dev/null  4.95s user 0.14s system 93% cpu 5.452 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.13s system 93% cpu 5.540 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.11s system 93% cpu 5.518 total
make -j1 -f posix.mak > /dev/null  5.06s user 0.14s system 93% cpu 5.565 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.12s system 93% cpu 5.503 total
make -j1 -f posix.mak > /dev/null  5.03s user 0.11s system 93% cpu 5.506 total
make -j1 -f posix.mak > /dev/null  4.97s user 0.15s system 93% cpu 5.491 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.13s system 93% cpu 5.501 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.11s system 93% cpu 5.508 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.12s system 93% cpu 5.495 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.12s system 93% cpu 5.474 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.12s system 93% cpu 5.521 total
make -j1 -f posix.mak > /dev/null  5.05s user 0.13s system 93% cpu 5.555 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.11s system 93% cpu 5.471 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.13s system 93% cpu 5.503 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.11s system 93% cpu 5.463 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.15s system 93% cpu 5.510 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.18s system 93% cpu 5.495 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.12s system 93% cpu 5.505 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.15s system 93% cpu 5.454 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.14s system 93% cpu 5.498 total
make -j1 -f posix.mak > /dev/null  4.95s user 0.12s system 93% cpu 5.446 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.17s system 93% cpu 5.529 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.16s system 93% cpu 5.507 total
make -j1 -f posix.mak > /dev/null  4.97s user 0.15s system 93% cpu 5.492 total
make -j1 -f posix.mak > /dev/null  5.03s user 0.14s system 93% cpu 5.532 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.14s system 93% cpu 5.511 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.18s system 93% cpu 5.485 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.14s system 93% cpu 5.492 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.12s system 93% cpu 5.501 total
make -j1 -f posix.mak > /dev/null  5.05s user 0.09s system 93% cpu 5.513 total
make -j1 -f posix.mak > /dev/null  5.06s user 0.14s system 93% cpu 5.562 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.16s system 93% cpu 5.554 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.12s system 93% cpu 5.528 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.11s system 93% cpu 5.477 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.18s system 93% cpu 5.532 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.13s system 93% cpu 5.536 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.11s system 93% cpu 5.476 total
make -j1 -f posix.mak > /dev/null  4.95s user 0.14s system 93% cpu 5.451 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.14s system 93% cpu 5.492 total
make -j1 -f posix.mak > /dev/null  5.07s user 0.14s system 93% cpu 5.589 total
make -j1 -f posix.mak > /dev/null  4.97s user 0.13s system 93% cpu 5.466 total
```

With this PR:

```
make -j1 -f posix.mak > /dev/null  5.17s user 0.14s system 93% cpu 5.679 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.10s system 93% cpu 5.505 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.14s system 93% cpu 5.492 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.15s system 93% cpu 5.453 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.18s system 93% cpu 5.551 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.16s system 93% cpu 5.511 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.14s system 93% cpu 5.477 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.14s system 93% cpu 5.515 total
make -j1 -f posix.mak > /dev/null  5.03s user 0.12s system 93% cpu 5.510 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.15s system 93% cpu 5.493 total
make -j1 -f posix.mak > /dev/null  4.93s user 0.18s system 93% cpu 5.476 total
make -j1 -f posix.mak > /dev/null  5.05s user 0.12s system 93% cpu 5.537 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.10s system 93% cpu 5.496 total
make -j1 -f posix.mak > /dev/null  5.03s user 0.11s system 93% cpu 5.509 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.12s system 93% cpu 5.486 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.18s system 93% cpu 5.585 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.13s system 93% cpu 5.487 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.15s system 93% cpu 5.482 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.12s system 93% cpu 5.456 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.14s system 93% cpu 5.506 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.15s system 93% cpu 5.472 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.14s system 93% cpu 5.500 total
make -j1 -f posix.mak > /dev/null  5.10s user 0.11s system 93% cpu 5.576 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.16s system 93% cpu 5.529 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.15s system 93% cpu 5.501 total
make -j1 -f posix.mak > /dev/null  4.99s user 0.10s system 93% cpu 5.465 total
make -j1 -f posix.mak > /dev/null  5.03s user 0.13s system 92% cpu 5.557 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.13s system 93% cpu 5.513 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.16s system 93% cpu 5.464 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.11s system 93% cpu 5.423 total
make -j1 -f posix.mak > /dev/null  5.04s user 0.14s system 93% cpu 5.561 total
make -j1 -f posix.mak > /dev/null  5.01s user 0.12s system 93% cpu 5.497 total
make -j1 -f posix.mak > /dev/null  5.06s user 0.13s system 93% cpu 5.559 total
make -j1 -f posix.mak > /dev/null  5.00s user 0.10s system 93% cpu 5.481 total
make -j1 -f posix.mak > /dev/null  5.07s user 0.12s system 93% cpu 5.559 total
make -j1 -f posix.mak > /dev/null  5.06s user 0.14s system 93% cpu 5.579 total
make -j1 -f posix.mak > /dev/null  4.94s user 0.16s system 93% cpu 5.464 total
make -j1 -f posix.mak > /dev/null  5.06s user 0.13s system 93% cpu 5.556 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.16s system 93% cpu 5.494 total
make -j1 -f posix.mak > /dev/null  4.93s user 0.16s system 93% cpu 5.459 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.13s system 93% cpu 5.481 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.11s system 93% cpu 5.498 total
make -j1 -f posix.mak > /dev/null  5.03s user 0.14s system 93% cpu 5.547 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.16s system 93% cpu 5.486 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.12s system 93% cpu 5.477 total
make -j1 -f posix.mak > /dev/null  5.08s user 0.10s system 93% cpu 5.555 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.10s system 93% cpu 5.489 total
make -j1 -f posix.mak > /dev/null  4.96s user 0.11s system 93% cpu 5.428 total
make -j1 -f posix.mak > /dev/null  4.98s user 0.14s system 93% cpu 5.488 total
make -j1 -f posix.mak > /dev/null  5.02s user 0.14s system 93% cpu 5.529 total
```

Minimum: 5.423 vs 5.446
Median: 5.497 vs 5.503

This PR is at worst as fast as master.

**Comment by andralex** [comment]
@WalterBright `typeToExpression` may be confusing, "how to you mean I convert a type to an expression"?

**Review by andralex — APPROVED** [review]
@ibuclaw @redstar @klickverbot @MartinNowak could you please take a look

**Comment by UplinkCoder** [comment]
@andralex aliasSeq works is a type. yet it can hold expressions.
this works via TypeExp.

File: src/ddmd/typesem.d [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        result = typeToExpressionHelper(t, new IdentifierExp(t.loc, t.ident));
+    }
+
+    override void visit(TypeInstance t)
+    {
+        result = typeToExpressionHelper(t, new ScopeExp(t.loc, t.tempinst));
+    }
+}
+
+extern (C++) Expression typeToExpression(Type t)
```
Comment (WalterBright) [inline src/ddmd/typesem.d]: No documentation.

File: src/ddmd/typesem.d [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+{
+    scope v = new TypeToExpressionVisitor();
+    t.accept(v);
+    return v.result;
+}
+
+extern (C++) Expression typeToExpressionHelper(TypeQualified t, Expression e, size_t i = 0)
```
Comment (WalterBright) [inline src/ddmd/typesem.d]: No documentation.

File: src/ddmd/initsem.d [hunk 3]
```diff
@@ -661,3 +662,211 @@ extern (C++) Initializer inferType(Initializer init, Scope* sc)
...
+                ae.type = itype;
+                result = ae;
+                return;
+            }
+        }
+        result = i.exp;
+    }
+}
+
+public extern (C++) Expression initializerToExpression(Initializer i, Type t = null)
```
Comment (WalterBright) [inline src/ddmd/initsem.d]: No documentation.

**Comment by RazvanN7** [comment]
@WalterBright I added the documentation for those functions.

**Comment by UplinkCoder** [comment]
@RazvanN7 This broke newCTFE.
now I have to use a less descriptive name then `toExpression`.

**Comment by UplinkCoder** [comment]
@RazvanN7 How do I get the `Initalizer.toExpression` ?
Please tell me how to migrate to the visitor base approuch.

Thanks!

**Comment by ZombineDev** [comment]
`initializerToExpression` ?

**Comment by RazvanN7** [comment]
@UplinkCoder As ZombineDev pointed out you have to use initializerToExpression for initializer nodes and typeToExpression for type nodes. I think these names are more expressive than toExpression.

**Pull request merged**